src/main/scala/com/databricks/labs/overwatch/pipeline/GoldTransforms.scala

```diff
@@ -493,6 +493,149 @@ trait GoldTransforms extends SparkSessionWrapper {
+
+    val clsfDF = clsfIncrementalDF
+      .select(
+        "organization_id", "state_start_date", "unixTimeMS_state_start", "cluster_id",
+        "current_num_workers", "uptime_in_state_H", "total_DBU_cost","unixTimeMS_state_end")
+      .distinct
+      //      .withColumnRenamed("unixTimeMS_state_start", "unixTimeMS")
+      .withColumnRenamed("cluster_id", "clusterId")
+      .withColumnRenamed("current_num_workers", "node_count")
+      .withColumn("total_dbu_cost_ps", col("total_DBU_cost") / col("uptime_in_state_H")/lit(3600))
+
+
+    val colNames: Array[Column] = Array(
+      'organization_id,
+      'clusterId.alias("cluster_id"),
+      'notebookId.alias("notebook_id"),
+      'workspace_name,
+      'date,
+      'timestamp,
+      'notebook_path,
+      'notebook_name,
+      'commandId.alias("command_id"),
+      'commandText.alias("command_text"),
+      'executionTime.alias("execution_time_s"),
+      'sourceIpAddress.alias("source_ip_address"),
+      'userIdentity.alias("user_identity"),
+      'estimated_dbu_cost,
+      'status,
+      'cluster_name,
+      'custom_tags,
+      'node_type_id,
+      'node_count,
+      'response,
+      'userAgent.alias("user_agent"),
+      'unixTimeMS
+    )
+    val auditDF = auditIncrementalDF
+      .filter(col("serviceName") === "notebook" && col("actionName") === "runCommand")
+      .selectExpr("*", "requestParams.*").drop("requestParams")
+      .withColumnRenamed("timestamp", "unixTimeMSStart")
+      .withColumn("timestamp", from_unixtime(col("unixTimeMSStart") / 1000.0).cast("timestamp"))
+      .withColumn("executionTime",col("executionTime").cast("double"))
+      .withColumn("unixTimeMSEnd",(col("unixTimeMSStart")+(col("executionTime")*1000)).cast("double"))
+      .filter('notebookId.isNotNull)
+    val joinedDF = auditDF.join(clsfDF,Seq("clusterId","organization_id"),"left")
+
+    // Cluster_state started before cmd start time and ended before command end time
+    val state_before_before = 'unixTimeMS_state_start < 'unixTimeMSStart && 'unixTimeMS_state_end < 'unixTimeMSEnd
+
+    // Cluster_state started on or after command start time and ended on or before command end time
+    val state_after_before = 'unixTimeMS_state_start >= 'unixTimeMSStart && 'unixTimeMS_state_end <= 'unixTimeMSEnd
+
+    // cluster_state started after command start time and ended after command end time
+    val state_after = 'unixTimeMS_state_start > 'unixTimeMSStart && 'unixTimeMS_state_end > 'unixTimeMSEnd
+
+    // Cluster State Started before cmd start time and ended after command run time
+    val state_before_after = 'unixTimeMS_state_start < 'unixTimeMSStart && 'unixTimeMS_state_end > 'unixTimeMSEnd
+
+    val windowSpec  = Window.partitionBy("commandId","clusterId","unixTimeMSStart","unixTimeMSEnd").orderBy("executionTime")
+
+    val joinedDF_before_before = joinedDF.filter(state_before_before)
+      .withColumn("avg_total_dbu_cost_ps", avg(col("total_dbu_cost_ps")).over(windowSpec))
```

Comment:
an average of averages will almost never yield the right numbers, and it will be incorrect in this case. Please surface the raw values from CLSF and do sums along the way, and only do an avg at the very end.